src/v/kafka/server/usage_aggregator.cc

```diff
@@ -0,0 +1,459 @@
+    std::optional<iobuf> period, windows, data;
+    try {
+        period = kvstore.get(kv_ks::usage, key_to_bytes(period_key));
+        windows = kvstore.get(kv_ks::usage, key_to_bytes(max_duration_key));
+        data = kvstore.get(kv_ks::usage, key_to_bytes(buckets_key));
+    } catch (const std::exception& ex) {
+        vlog(
+          klog.debug,
+          "Encountered exception when retriving usage data from disk: {}",
+          ex);
+        return std::nullopt;
+    }
+    if (!period && !windows && !data) {
+        /// Data didn't exist
+        return std::nullopt;
+    } else if (!period || !windows || !data) {
+        vlog(
+          klog.error,
+          "Inconsistent usage_manager on disk state detected, failed to "
+          "recover state");
+        return std::nullopt;
+    }
+    return persisted_state{
+      .configured_period = serde::from_iobuf<std::chrono::seconds>(
+        std::move(*period)),
+      .configured_windows = serde::from_iobuf<size_t>(std::move(*windows)),
+      .current_state = serde::from_iobuf<fragmented_vector<usage_window>>(
+        std::move(*data))};
+}
+
+static ss::future<> clear_persisted_state(storage::kvstore& kvstore) {
+    using kv_ks = storage::kvstore::key_space;
+    try {
+        co_await kvstore.remove(kv_ks::usage, key_to_bytes(period_key));
+        co_await kvstore.remove(kv_ks::usage, key_to_bytes(max_duration_key));
+        co_await kvstore.remove(kv_ks::usage, key_to_bytes(buckets_key));
+    } catch (const std::exception& ex) {
+        vlog(klog.debug, "Ignoring exception from storage layer: {}", ex);
+    }
+}
+
+usage usage::operator+(const usage& other) const {
+    return usage{
+      .bytes_sent = bytes_sent + other.bytes_sent,
+      .bytes_received = bytes_received + other.bytes_received};
+}
+
+usage& usage::operator+=(const usage& other) {
+    if (this != &other) {
+        bytes_sent += other.bytes_sent;
+        bytes_received += other.bytes_received;
+        bytes_cloud_storage = other.bytes_cloud_storage;
+    }
+    return *this;
+}
+
+template<typename time_point>
+auto epoch_time_secs(time_point now) {
+    return std::chrono::duration_cast<std::chrono::seconds>(
+             now.time_since_epoch())
+      .count();
+}
+
+static ss::lowres_system_clock::time_point round_to_interval(
+  std::chrono::seconds usage_window_width_interval,
+  ss::lowres_system_clock::time_point t) {
+    /// Downstream systems are particularly sensitive to minor issues with
+    /// timestamps not triggering on the configured interval (hours, minutes,
+    /// seconds, etc), this method rounds t to the nearest interval and logs an
+    /// error if this cannot be done within some threshold.
+    static const auto err_threshold = 5s;
+    const auto interval = usage_window_width_interval;
+    const auto cur_interval_start = t - (t.time_since_epoch() % interval);
+    const auto next_interval_start = cur_interval_start + interval;
```

Comment:
this doesn't seem to handle the case where t is slightly under the perfectly aligned time. for example, if t is 3599 then the start time rounds down to the previous hour boundary, and next start time will be 1 second in the future, 3600. but that probably isn't what the intention would be because it would mean you're effectively repeating the previous window?

it seems like what we want to do is round to the nearest hour if we are below or above the top of the hour?

Reply:
`t` could never be below 3599 though because its an epoch timestamp , coming from the system clock, but i think we could still handle that situation with some other maths

Reply:
I originally had this but modified it due to concerns over readability:
```
    static const auto err_threshold = 5s;
    const auto int_secs = usage_window_width_interval.count();
    const auto diff_secs = std::chrono::seconds(epoch_time_secs(t) % int_secs);
    const auto interval_lb = diff_secs
                             > (usage_window_width_interval - err_threshold);
    const auto interval_ub = diff_secs < err_threshold;
    if (!interval_lb && !interval_ub) {
        vlog(
          klog.error,
          .....
 ```